cli/parameterresolver.go

```diff
@@ -0,0 +1,224 @@
+package cli
+
+import (
+	"fmt"
+
+	"golang.org/x/xerrors"
+
+	"github.com/coder/coder/cli/clibase"
+	"github.com/coder/coder/cli/cliui"
+	"github.com/coder/coder/codersdk"
+)
+
+type WorkspaceCLIAction int
+
+const (
+	WorkspaceCreate WorkspaceCLIAction = iota
+	WorkspaceStart
+	WorkspaceUpdate
+	WorkspaceRestart
+)
+
+type ParameterResolver struct {
+	lastBuildParameters []codersdk.WorkspaceBuildParameter
+
+	richParameters     []codersdk.WorkspaceBuildParameter
+	richParametersFile map[string]string
+	buildOptions       []codersdk.WorkspaceBuildParameter
+
+	promptRichParameters bool
+	promptBuildOptions   bool
+}
+
+func (pr *ParameterResolver) WithLastBuildParameters(params []codersdk.WorkspaceBuildParameter) *ParameterResolver {
+	pr.lastBuildParameters = params
+	return pr
+}
+
+func (pr *ParameterResolver) WithRichParameters(params []codersdk.WorkspaceBuildParameter) *ParameterResolver {
+	pr.richParameters = params
+	return pr
+}
+
+func (pr *ParameterResolver) WithBuildOptions(params []codersdk.WorkspaceBuildParameter) *ParameterResolver {
+	pr.buildOptions = params
+	return pr
+}
+
+func (pr *ParameterResolver) WithRichParametersFile(fileMap map[string]string) *ParameterResolver {
+	pr.richParametersFile = fileMap
+	return pr
+}
+
+func (pr *ParameterResolver) WithPromptRichParameters(promptRichParameters bool) *ParameterResolver {
+	pr.promptRichParameters = promptRichParameters
+	return pr
+}
+
+func (pr *ParameterResolver) WithPromptBuildOptions(promptBuildOptions bool) *ParameterResolver {
+	pr.promptBuildOptions = promptBuildOptions
+	return pr
+}
+
+func (pr *ParameterResolver) Resolve(inv *clibase.Invocation, action WorkspaceCLIAction, templateVersionParameters []codersdk.TemplateVersionParameter) ([]codersdk.WorkspaceBuildParameter, error) {
+	var staged []codersdk.WorkspaceBuildParameter
+	var err error
+
+	staged = pr.resolveWithParametersMapFile(staged)
+	staged = pr.resolveWithCommandLineOrEnv(staged)
+	staged = pr.resolveWithLastBuildParameters(staged, templateVersionParameters)
+	if err = pr.verifyConstraints(staged, action, templateVersionParameters); err != nil {
+		return nil, err
+	}
+	if staged, err = pr.resolveWithInput(staged, inv, action, templateVersionParameters); err != nil {
+		return nil, err
+	}
+	return staged, nil
+}
+
+func (pr *ParameterResolver) resolveWithParametersMapFile(resolved []codersdk.WorkspaceBuildParameter) []codersdk.WorkspaceBuildParameter {
+	for name, value := range pr.richParametersFile {
+		for i, r := range resolved {
+			if r.Name == name {
+				resolved[i].Value = value
+				goto done
```

Comment:
Suggestion: We could label the loop instead and use `continue Loop`. Continue/break have a simpler flow and are usually considered more idiomatic Go.

(Same with other loops.)

Reply:
Are you thinking about this form:

```go
next:
	for name, value := range pr.richParametersFile {
		for i, r := range resolved {
			if r.Name == name {
				resolved[i].Value = value
				continue next
			}
		}

		resolved = append(resolved, codersdk.WorkspaceBuildParameter{
			Name:  name,
			Value: value,
		})
	}
```

If so, then it will be a bit different logic - iterate over `pr.richParametersFile` again and again.

Alternatively, I can add break/if, but with `goto` the code is shorter/simpler.

Reply:
I think there's a misunderstanding around how loop labels work. This is kind of like tagging a loop so that `continue` knows where to go. See here (only one iteration through the range is performed): https://go.dev/play/p/oAbzFLuh8O5

Reply:
https://github.com/coder/coder/pull/9019